src/pip/_internal/commands/install.py

```diff
@@ -321,21 +315,51 @@ def run(self, options, args):
                         modifying_pip=requirement_set.has_requirement("pip")
                     )
 
-                    # If caching is disabled or wheel is not installed don't
-                    # try to build wheels.
-                    if wheel and options.cache_dir:
-                        # build wheels before install.
-                        wb = WheelBuilder(
-                            finder, preparer, wheel_cache,
-                            build_options=[], global_options=[],
-                        )
-                        # Ignore the result: a failed wheel will be
-                        # installed from the sdist/vcs whatever.
+                    # Consider legacy and PEP517-using requirements separately
+                    legacy_requirements = []
+                    pep517_requirements = []
+                    for req in requirement_set.requirements.values():
+                        if req.use_pep517:
+                            pep517_requirements.append(req)
+                        else:
+                            legacy_requirements.append(req)
+
+                    # We don't build wheels for legacy requirements if we
+                    # don't have wheel installed or we don't have a cache dir
+                    try:
+                        import wheel  # noqa: F401
+                        build_legacy = bool(options.cache_dir)
+                    except ImportError:
+                        build_legacy = False
+
+                    wb = WheelBuilder(
+                        finder, preparer, wheel_cache,
+                        build_options=[], global_options=[],
+                    )
+
+                    # Always build PEP 517 requirements
+                    build_failures = wb.build(
+                        pep517_requirements,
+                        session=session, autobuilding=True
+                    )
+
+                    if build_legacy:
+                        # We don't care about failures building legacy
+                        # requirements, as we'll fall through to a direct
+                        # install for those.
                         wb.build(
-                            requirement_set.requirements.values(),
+                            pep517_requirements,
```

Comment:
Shouldn't this be `legacy_requirements`? I think this is the source of all the test failures.

That leaves us with only one more failure -- `test_install_no_binary_disables_building_wheels`.

Reply:
Ouch, that was a stupid copy/paste error on my part. Thanks for catching this.

Reply:
And `test_install_no_binary_disables_building_wheels` was shallow, too, I'd messed up modifying the test to cater for pip's changed output. A lot of the tests depend on the precise wording of pip's output for their checks, rather than checking the actual behaviour directly. So I had to do some editing of tests just to match changed output (I've changed a lot of messages to not refer to `setup.py` but say something more generic like "Building wheel"). And I messed one such change up.

As it stood, the test was checking *both* `"Building wheel for requir" in str(res)` and `"Building wheel for requir" not in str(res)`. If it had ever passed, I'd have invented a new form of mathematical logic :-)